### PR TITLE
Fix filtering of dictionaries with tuple keys

### LINKED ISSUE #165: 500 Error created by Honeybadger send_notice
I think this is probably a honeybadger edge-case bug. 

There is an endpoint which I’m hitting badly intentionally, it should return a 422 because I’m passing 1234 instead of a valid guid. 

It works well in my local test environment because hb is disabled in local, but in production, where honey badger is enabled it results in a 500 error, and no notice is sent to hb, because it is failing in the notification sending itself. This is the error trace I see in the server:
```
ERROR:    Exception in ASGI application
agiliza-web-1      | Traceback (most recent call last):
agiliza-web-1      |   File "/usr/local/lib/python3.11/site-packages/honeybadger/contrib/asgi.py", line 115, in _run_app
agiliza-web-1      |     return await callback()
agiliza-web-1      |            ^^^^^^^^^^^^^^^^
agiliza-web-1      |   File "/usr/local/lib/python3.11/site-packages/starlette/middleware/base.py", line 189, in __call__
agiliza-web-1      |     with collapse_excgroups():
agiliza-web-1      |   File "/usr/local/lib/python3.11/contextlib.py", line 155, in __exit__
agiliza-web-1      |     self.gen.throw(typ, value, traceback)
agiliza-web-1      |   File "/usr/local/lib/python3.11/site-packages/starlette/_utils.py", line 91, in collapse_excgroups
agiliza-web-1      |     raise exc
agiliza-web-1      |   File "/usr/local/lib/python3.11/site-packages/starlette/middleware/base.py", line 191, in __call__
agiliza-web-1      |     response = await self.dispatch_func(request, call_next)
agiliza-web-1      |                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
agiliza-web-1      |   File "/usr/local/lib/python3.11/site-packages/starlette_prometheus/middleware.py", line 57, in dispatch
agiliza-web-1      |     raise e from None
agiliza-web-1      |   File "/usr/local/lib/python3.11/site-packages/starlette_prometheus/middleware.py", line 53, in dispatch
agiliza-web-1      |     response = await call_next(request)
agiliza-web-1      |                ^^^^^^^^^^^^^^^^^^^^^^^^
agiliza-web-1      |   File "/usr/local/lib/python3.11/site-packages/starlette/middleware/base.py", line 165, in call_next
agiliza-web-1      |     raise app_exc
agiliza-web-1      |   File "/usr/local/lib/python3.11/site-packages/starlette/middleware/base.py", line 151, in coro
agiliza-web-1      |     await self.app(scope, receive_or_disconnect, send_no_error)
agiliza-web-1      |   File "/usr/local/lib/python3.11/site-packages/starlette/middleware/authentication.py", line 48, in __call__
agiliza-web-1      |     await self.app(scope, receive, send)
agiliza-web-1      |   File "/usr/local/lib/python3.11/site-packages/starlette/middleware/cors.py", line 83, in __call__
agiliza-web-1      |     await self.app(scope, receive, send)
agiliza-web-1      |   File "/usr/local/lib/python3.11/site-packages/starlette/middleware/exceptions.py", line 62, in __call__
agiliza-web-1      |     await wrap_app_handling_exceptions(self.app, conn)(scope, receive, send)
agiliza-web-1      |   File "/usr/local/lib/python3.11/site-packages/starlette/_exception_handler.py", line 64, in wrapped_app
agiliza-web-1      |     raise exc
agiliza-web-1      |   File "/usr/local/lib/python3.11/site-packages/starlette/_exception_handler.py", line 53, in wrapped_app
agiliza-web-1      |     await app(scope, receive, sender)
agiliza-web-1      |   File "/usr/local/lib/python3.11/site-packages/starlette/routing.py", line 762, in __call__
agiliza-web-1      |     await self.middleware_stack(scope, receive, send)
agiliza-web-1      |   File "/usr/local/lib/python3.11/site-packages/starlette/routing.py", line 782, in app
agiliza-web-1      |     await route.handle(scope, receive, send)
agiliza-web-1      |   File "/usr/local/lib/python3.11/site-packages/starlette/routing.py", line 297, in handle
agiliza-web-1      |     await self.app(scope, receive, send)
agiliza-web-1      |   File "/usr/local/lib/python3.11/site-packages/starlette/routing.py", line 77, in app
agiliza-web-1      |     await wrap_app_handling_exceptions(app, request)(scope, receive, send)
agiliza-web-1      |   File "/usr/local/lib/python3.11/site-packages/starlette/_exception_handler.py", line 64, in wrapped_app
agiliza-web-1      |     raise exc
agiliza-web-1      |   File "/usr/local/lib/python3.11/site-packages/starlette/_exception_handler.py", line 53, in wrapped_app
agiliza-web-1      |     await app(scope, receive, sender)
agiliza-web-1      |   File "/usr/local/lib/python3.11/site-packages/starlette/routing.py", line 72, in app
agiliza-web-1      |     response = await func(request)
agiliza-web-1      |                ^^^^^^^^^^^^^^^^^^^
agiliza-web-1      |   File "/usr/local/lib/python3.11/site-packages/honeybadger/contrib/fastapi.py", line 25, in custom_route_handler
agiliza-web-1      |     honeybadger.notify(exception=exc, context=asgi._as_context(scope))
agiliza-web-1      |   File "/usr/local/lib/python3.11/site-packages/honeybadger/core.py", line 57, in notify
agiliza-web-1      |     return self._send_notice(exception, context=merged_context, fingerprint=fingerprint)
agiliza-web-1      |            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
agiliza-web-1      |   File "/usr/local/lib/python3.11/site-packages/honeybadger/core.py", line 27, in _send_notice
agiliza-web-1      |     return connection.send_notice(self.config, payload)
agiliza-web-1      |            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
agiliza-web-1      |   File "/usr/local/lib/python3.11/site-packages/honeybadger/connection.py", line 16, in send_notice
agiliza-web-1      |     data=b(json.dumps(payload, cls=StringReprJSONEncoder)))
agiliza-web-1      |            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
agiliza-web-1      |   File "/usr/local/lib/python3.11/json/__init__.py", line 238, in dumps
agiliza-web-1      |     **kw).encode(obj)
agiliza-web-1      |           ^^^^^^^^^^^
agiliza-web-1      |   File "/usr/local/lib/python3.11/json/encoder.py", line 200, in encode
agiliza-web-1      |     chunks = self.iterencode(o, _one_shot=True)
agiliza-web-1      |              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
agiliza-web-1      |   File "/usr/local/lib/python3.11/json/encoder.py", line 258, in iterencode
agiliza-web-1      |     return _iterencode(o, 0)
agiliza-web-1      |            ^^^^^^^^^^^^^^^^^
agiliza-web-1      | TypeError: keys must be str, int, float, bool or None, not tuple
agiliza-web-1      |
agiliza-web-1      | During handling of the above exception, another exception occurred:
agiliza-web-1      |
agiliza-web-1      | Traceback (most recent call last):
agiliza-web-1      |   File "/usr/local/lib/python3.11/site-packages/uvicorn/protocols/http/httptools_impl.py", line 419, in run_asgi
agiliza-web-1      |     result = await app(  # type: ignore[func-returns-value]
agiliza-web-1      |              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
agiliza-web-1      |   File "/usr/local/lib/python3.11/site-packages/uvicorn/middleware/proxy_headers.py", line 84, in __call__
agiliza-web-1      |     return await self.app(scope, receive, send)
agiliza-web-1      |            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
agiliza-web-1      |   File "/usr/local/lib/python3.11/site-packages/fastapi/applications.py", line 1054, in __call__
agiliza-web-1      |     await super().__call__(scope, receive, send)
agiliza-web-1      |   File "/usr/local/lib/python3.11/site-packages/starlette/applications.py", line 123, in __call__
agiliza-web-1      |     await self.middleware_stack(scope, receive, send)
agiliza-web-1      |   File "/usr/local/lib/python3.11/site-packages/starlette/middleware/errors.py", line 186, in __call__
agiliza-web-1      |     raise exc
agiliza-web-1      |   File "/usr/local/lib/python3.11/site-packages/starlette/middleware/errors.py", line 164, in __call__
agiliza-web-1      |     await self.app(scope, receive, _send)
agiliza-web-1      |   File "/usr/local/lib/python3.11/site-packages/honeybadger/contrib/asgi.py", line 109, in _run_asgi3
agiliza-web-1      |     return await self._run_app(scope, lambda: self.app(scope, receive, send))
agiliza-web-1      |            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
agiliza-web-1      |   File "/usr/local/lib/python3.11/site-packages/honeybadger/contrib/asgi.py", line 117, in _run_app
agiliza-web-1      |     honeybadger.notify(exception=exc, context=_as_context(scope))
agiliza-web-1      |   File "/usr/local/lib/python3.11/site-packages/honeybadger/core.py", line 57, in notify
agiliza-web-1      |     return self._send_notice(exception, context=merged_context, fingerprint=fingerprint)
agiliza-web-1      |            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
agiliza-web-1      |   File "/usr/local/lib/python3.11/site-packages/honeybadger/core.py", line 27, in _send_notice
agiliza-web-1      |     return connection.send_notice(self.config, payload)
agiliza-web-1      |            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
agiliza-web-1      |   File "/usr/local/lib/python3.11/site-packages/honeybadger/connection.py", line 16, in send_notice
agiliza-web-1      |     data=b(json.dumps(payload, cls=StringReprJSONEncoder)))
agiliza-web-1      |            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
agiliza-web-1      |   File "/usr/local/lib/python3.11/json/__init__.py", line 238, in dumps
agiliza-web-1      |     **kw).encode(obj)
agiliza-web-1      |           ^^^^^^^^^^^
agiliza-web-1      |   File "/usr/local/lib/python3.11/json/encoder.py", line 200, in encode
agiliza-web-1      |     chunks = self.iterencode(o, _one_shot=True)
agiliza-web-1      |              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
agiliza-web-1      |   File "/usr/local/lib/python3.11/json/encoder.py", line 258, in iterencode
agiliza-web-1      |     return _iterencode(o, 0)
agiliza-web-1      |            ^^^^^^^^^^^^^^^^^
agiliza-web-1      | TypeError: keys must be str, int, float, bool or None, not tuple
```
I can perfectly reproduce this in my local environment. I have been having a look, but I’m a bit lost at what is happening. I’ve narrowed down the last error and I see that indeed there is a dict with a key which is hard to encode.

What I see is that finally, at encoder.py, line 200 (actually seems to be line 199 for me, at

`chunks = self.iterencode(o, _one_shot=True)`

I see that it fails if I try `self.iterencode(o["request"]["local_variables"]["_"], _one_shot=True)`

This ‘_’ key looks like this:
image

Why this appears in the `local_variables`, of course I have no clue.

So, then this raises an exception inside honey badger trying to send the notification to the service, which explains why no error appears in hb itself. If I disable hb, I get the expected 422 response.

In payload.py, line 118:
image

I can see that the local_variables do indeed have a `'_'` variable in it:

image

I have worked around this by disabling local variable collection in honeybadger config in the meanwhile until a fix is found.

### PR DESCRIPTION
**Pull request by Kelvin4664**
Resolves #165

### CODE REVIEW

File: honeybadger/utils.py [hunk 1]
```diff
@@ -14,11 +15,20 @@ def filter_dict(data, filter_keys):
     if type(data) != dict:
         return data
 
+    data_copy = copy.deepcopy(data)
+
     for key, value in data.items():
```
Comment (subzero10) [inline honeybadger/utils.py:20]: For better safety, I'd probably iterate over `data_copy` here.

**Review by subzero10 — APPROVED** [review]
Looks good, thanks! Just a minor comment.
